query/update/operator.go

```diff
@@ -57,30 +60,52 @@ type FieldOperatorFactory struct {
 	FieldOperators map[string]*FieldOperator
 }
 
-// MergeAndGet method to converts the input to the output after applying all the operators.
+// MergeAndGet method to converts the input to the output after applying all the operators. First "$set" operation is
+// applied and then "$unset" which means if a field is present in both $set and $unset then it won't be stored in the
+// resulting document.
 func (factory *FieldOperatorFactory) MergeAndGet(existingDoc jsoniter.RawMessage) (jsoniter.RawMessage, error) {
-	setFieldOp := factory.FieldOperators[string(Set)]
-	if setFieldOp == nil {
-		return nil, errors.InvalidArgument("set operator not present in the fields parameter")
+	out := existingDoc
+	var err error
+	if setFieldOp, ok := factory.FieldOperators[string(Set)]; ok {
+		if out, err = factory.set(out, setFieldOp.Input); err != nil {
+			return nil, err
+		}
 	}
-	out, err := factory.apply(existingDoc, setFieldOp.Document)
-	if err != nil {
+	if unsetFieldOp, ok := factory.FieldOperators[string(UnSet)]; ok {
+		if out, err = factory.remove(out, unsetFieldOp.Input); err != nil {
+			return nil, err
+		}
+	}
+
+	return out, nil
+}
+
+func (factory *FieldOperatorFactory) remove(out jsoniter.RawMessage, toRemove jsoniter.RawMessage) (jsoniter.RawMessage, error) {
+	var unsetArray []string
+	if err := jsoniter.Unmarshal(toRemove, &unsetArray); err != nil {
 		return nil, err
 	}
 
+	for _, unset := range unsetArray {
+		unsetKeys := strings.Split(unset, ".")
+		out = jsonparser.Delete(out, unsetKeys...)
+	}
+
 	return out, nil
 }
 
-func (factory *FieldOperatorFactory) apply(input jsoniter.RawMessage, setDoc jsoniter.RawMessage) (jsoniter.RawMessage, error) {
+func (factory *FieldOperatorFactory) set(existingDoc jsoniter.RawMessage, setDoc jsoniter.RawMessage) (jsoniter.RawMessage, error) {
 	var (
-		output []byte = input
+		output []byte = existingDoc
 		err    error
 	)
 	err = jsonparser.ObjectEach(setDoc, func(key []byte, value []byte, dataType jsonparser.ValueType, offset int) error {
 		if dataType == jsonparser.String {
 			value = []byte(fmt.Sprintf(`"%s"`, value))
 		}
-		output, err = jsonparser.Set(output, value, string(key))
+
+		keys := strings.Split(string(key), ".")
```

Comment:
Should we have a helper for this?
Also, is this how we specify nested fields? "a.b.c" ?

Reply:
The helper will also be a single method, not sure if it would be beneficial here. Isn't `a.b.c` the current client semantics of sending the nested fields?

Reply:
` Isn't a.b.c the current client semantics of sending the nested fields?`
I just didn't notice this nested semantic in the tests, now I see it.